node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -314,7 +318,11 @@ handleCHsValid enqueue endedRecoveryVar nodeId lcaChild newestHash = do
                                 then isJust <$> tryTakeTMVar recHeaderVar
                                 else return False
                 void $ tryPutMVar endedRecoveryVar endedRecovery
-    void $ waitForConversations its
+    -- Try to fill the MVar at the end. In case the conversation is run
+    -- and succesfully ends it, it will already be filled. That's fine.
+    -- But in case there's an error retrieving the blocks, or in case the
+    -- conversation is enqueued to nobody, we have to ensure the MVar is filled.
+    void $ waitForConversations its `finally` tryPutMVar endedRecoveryVar False
```

Comment:
> So the point here is that waitForConversations is the bit that throws exceptions for conversations that failed right?

That's true.

> So are we sure in that case we will be ending recovery?

Maybe not. It won't happen in practice, because we never enqueue to more than 1 peer, but it *could* happen that there's >1 conversation and one of them throws an exception before the other completes normally. In that case the MVar should have `True` but will have `False`.